test.js

```diff
@@ -316,3 +316,112 @@ test('missing values', (t) => {
   t.equal('{"str":"string","val":"value"}', stringify({ str: 'string', val: 'value' }))
   t.equal('{"str":"string","num":42,"val":"value"}', stringify({ str: 'string', num: 42, val: 'value' }))
 })
+
+test('patternProperties', (t) => {
+  t.plan(7)
+  let stringify = build({
+    title: 'patternProperties',
+    type: 'object',
+    properties: {
+      str: {
+        type: 'string'
+      }
+    },
+    patternProperties: {
+      'foo': {
+        type: 'string'
+      }
+    }
+  })
+
+  let obj = { str: 'test', foo: 42, ofoo: true, foof: 'string', objfoo: {a: true}, notMe: false }
+  t.equal('{"foo":"42","ofoo":"true","foof":"string","objfoo":"[object Object]","str":"test"}', stringify(obj))
+
+  stringify = build({
+    title: 'patternProperties should not change properties',
+    type: 'object',
+    properties: {
+      foo: {
+        type: 'string'
+      }
+    },
+    patternProperties: {
+      foo: {
+        type: 'number'
+      }
+    }
+  })
+
+  obj = { foo: '42', ofoo: 42 }
+  t.equal('{"ofoo":42,"foo":"42"}', stringify(obj))
+
+  stringify = build({
+    title: 'check string coerce',
+    type: 'object',
+    properties: {},
+    patternProperties: {
+      foo: {
+        type: 'string'
+      }
+    }
+  })
+
+  obj = { foo: true, ofoo: 42, arrfoo: ['array', 'test'], objfoo: { a: 'world' } }
+  t.equal('{"foo":"true","ofoo":"42","arrfoo":"array,test","objfoo":"[object Object]"}', stringify(obj))
+
+  stringify = build({
+    title: 'check number coerce',
+    type: 'object',
+    properties: {},
+    patternProperties: {
+      foo: {
+        type: 'number'
+      }
+    }
+  })
+
+  obj = { foo: true, ofoo: '42', xfoo: 'string', arrfoo: [1, 2], objfoo: { num: 42 } }
+  t.equal('{"foo":1,"ofoo":42,"xfoo":null,"arrfoo":null,"objfoo":null}', stringify(obj))
+
+  stringify = build({
+    title: 'check boolean coerce',
+    type: 'object',
+    properties: {},
+    patternProperties: {
+      foo: {
+        type: 'boolean'
+      }
+    }
+  })
+
+  obj = { foo: 'true', ofoo: 0, arrfoo: [1, 2], objfoo: { a: true } }
+  t.equal('{"foo":true,"ofoo":false,"arrfoo":true,"objfoo":true}', stringify(obj))
+
+  stringify = build({
+    title: 'check object coerce',
+    type: 'object',
+    properties: {},
+    patternProperties: {
+      foo: {
+        type: 'object'
+      }
+    }
+  })
+
+  obj = { foo: true, ofoo: '42', arrfoo: [1, 2], objfoo: { answer: 42 } }
+  t.equal('{"foo":{},"ofoo":{},"arrfoo":{},"objfoo":{}}', stringify(obj))
+
+  stringify = build({
+    title: 'check array coerce',
+    type: 'object',
+    properties: {},
+    patternProperties: {
+      foo: {
+        type: 'array'
+      }
+    }
+  })
+
+  obj = { foo: 'true', ofoo: 0, arrfoo: [1, 2], objfoo: { tyrion: 'lannister' } }
+  t.equal('{"foo":[],"ofoo":[],"arrfoo":[],"objfoo":[]}', stringify(obj))
```

Comment:
You can split each of those tests in its own `test`. They are hard to read.
You can also split each feature in its own file.
See https://github.com/mcollina/pino as an example.